internal/worker/asynccharmdownloader/downloader.go

```diff
@@ -0,0 +1,121 @@
+// Copyright 2024 Canonical Ltd.
+// Licensed under the AGPLv3, see LICENCE file for details.
+
+package asynccharmdownloader
+
+import (
+	"context"
+	"time"
+
+	"github.com/juju/clock"
+	"github.com/juju/retry"
+	"github.com/juju/worker/v4"
+	"gopkg.in/tomb.v2"
+
+	"github.com/juju/juju/core/application"
+	"github.com/juju/juju/core/logger"
+	domainapplication "github.com/juju/juju/domain/application"
+	applicationerrors "github.com/juju/juju/domain/application/errors"
+	"github.com/juju/juju/internal/charm/charmdownloader"
+	"github.com/juju/juju/internal/errors"
+)
+
+const (
+	retryAttempts = 3
+	retryDelay    = 20 * time.Second
+)
+
+type asyncDownloadWorker struct {
+	tomb tomb.Tomb
+
+	appID              application.ID
+	applicationService ApplicationService
+	downloader         Downloader
+
+	clock  clock.Clock
+	logger logger.Logger
+}
+
+// NewAsyncDownloadWorker creates a new async worker that downloads charms for
+// the specified application.
+func NewAsyncDownloadWorker(
+	appID application.ID,
+	applicationService ApplicationService,
+	downloader Downloader,
+	clock clock.Clock,
+	logger logger.Logger,
+) worker.Worker {
+	w := &asyncDownloadWorker{
+		appID:              appID,
+		applicationService: applicationService,
+		downloader:         downloader,
+		clock:              clock,
+		logger:             logger,
+	}
+	w.tomb.Go(w.loop)
+	return w
+}
+
+// Kill is part of the worker.Worker interface.
+func (w *asyncDownloadWorker) Kill() {
+	w.tomb.Kill(nil)
+}
+
+// Wait is part of the worker.Worker interface.
+func (w *asyncDownloadWorker) Wait() error {
+	return w.tomb.Wait()
+}
+
+func (w *asyncDownloadWorker) loop() error {
+	ctx, cancel := w.scopedContext()
+	defer cancel()
+
+	w.logger.Infof("downloading charm for application %q", w.appID)
+
+	info, err := w.applicationService.ReserveCharmDownload(ctx, w.appID)
```

Comment:
OK, I think the naming should be reviewed because (I think) you intended these `Reserve` & `Resolve` almost as locks. Maybe they should have very very explicit names like `Lock` & `Unlock`?

Reply:
Agreed, these aren't the best names for these, but we'll be changing the service layer, so we'll rename then.

Reply:
Happy with the current names.